src/coverage/coverageService.ts

```diff
@@ -0,0 +1,163 @@
+// Copyright 2019 Espressif Systems (Shanghai) CO LTD
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { basename, join } from "path";
+import * as vscode from "vscode";
+import { appendIdfAndToolsToPath, execChildProcess } from "../utils";
+import { OutputChannel } from "../logger/outputChannel";
+import { Logger } from "../logger/logger";
+
+export interface countRange {
+  range: vscode.Range;
+  count: Number;
+}
+
+export interface textEditorWithCoverage {
+  coveredLines: vscode.Range[];
+  partialLines: vscode.Range[];
+  uncoveredLines: vscode.Range[];
+  editor: vscode.TextEditor;
+  countPerLines: countRange[];
+  allLines: vscode.Range[];
+}
+
+export async function buildJson(dirPath: string) {
+  const result = await _runCmd(
+    "gcovr",
+    [
+      "--filter",
+      ".",
+      "--filter",
+      "/Users/brian/esp-idf/components/",
+      "--gcov-executable",
+      "xtensa-esp32-elf-gcov",
+      "--json",
+    ],
+    dirPath
+  );
+  return JSON.parse(result);
+}
+
+export async function buildHtml(dirPath: string) {
+  const result = await _runCmd(
+    "gcovr",
+    [
+      "--filter",
+      ".",
+      "--filter",
+      "/Users/brian/esp-idf/components/",
+      "--gcov-executable",
+      "xtensa-esp32-elf-gcov",
+      "--html",
+    ],
+    dirPath
+  );
+  return result;
+}
+
+function _runCmd(cmd: string, args: string[], dirPath: string) {
+  const modifiedEnv = appendIdfAndToolsToPath();
+  return execChildProcess(
+    `${cmd} ${args.join(" ")}`,
+    dirPath,
+    OutputChannel.init(),
+    { env: modifiedEnv, cwd: dirPath }
+  ).catch((e) => {
+    const msg = e.message ? e.message : e;
+    Logger.error("Error on gcov cmd.\n" + msg, e);
+    OutputChannel.appendLine("Error building gcov cmd.\n" + msg);
+    return "";
+  });
+}
+
+export async function generateCoverageForEditors(
+  editors: vscode.TextEditor[],
+  dirPath: string
+) {
+  const gcovObj = await buildJson(dirPath);
```

Comment:
__Some nitpicking:__
For the scenario where, say you have total 5 source file and you have currently 2 of them opened.
Then you run the `Add Editor Coverage` command, it will render coverage report on the editor for 2 of those already opened source files.
But then if you open 3rd source file or close the already opened one and reopen that.
__Rendering of coverage line on the editor will be gone__ to solve this we have to re run the command, which will call `gcovr` again and again for re computing the same coverage report again.

__Suggestion:__
Maybe we can compute the coverage once for a given fileset and cache (and if that changes then only re compute like how idf build works), and `Add Editor Coverage ` will set a global flag and whenever a new editor tab opens it should check the flag and if the coverage for the source exists then paint the coverage data onto the editor!

Reply:
Also navigating between already opened editors is also loosing the coverage data highlights

Reply:
Yes, this a standard vscode behavior, decorations disappear when editors are hidden.

Will implement a cache system that re-render on `vscode.window. onDidChangeVisibleTextEditors()` event.